rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp

```diff
@@ -0,0 +1,60 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_LIFECYCLE__CHANGE_STATE_HANDLER_HPP_
+#define RCLCPP_LIFECYCLE__CHANGE_STATE_HANDLER_HPP_
+
+#include "rclcpp_lifecycle/node_interfaces/lifecycle_node_interface.hpp"
+
+namespace rclcpp_lifecycle
+{
+/// The object passed to asynchronous change_state user transition functions
+class ChangeStateHandler
+{
+public:
+  /// Continues the change state process handling proper callback order
+  /** Used within the user defined transition callback to continue the change state process
+   *  similar to a service call response
+   *  Note this only allows sending a single response callback per object
+   *  and will not send further responses if called mutiple times on the object
+   * \param[in] cb_return_code result of user defined transition callback
+   * \return true if the response was successfully sent
+   */
+  virtual bool send_callback_resp(
+    node_interfaces::LifecycleNodeInterface::CallbackReturn cb_return_code) = 0;
+
+  /// Updates the state machine based on the handling of a cancelled transition
+  /**
+   * \param[in] success true if the transition cancel request was successfully handled
+   * \return true if the response was successfully sent to the state handler
+  */
+  virtual bool canceled(bool success) = 0;
```

Comment:
the name of this method seems like we are just checking the state with no side-effects...
what about a different name such as `handle_canceled` ?

Reply:
This is essentially what it was initially called (and I prefer from a readability standpoint). 

However, I decided to follow the same exact conventions of ROS 2 Action which has a `canceled` [here](https://github.com/ros2/rclcpp/blob/rolling/rclcpp_action/include/rclcpp_action/server_goal_handle.hpp#L208) with a response (which is the analog to this function). Example of this being called [here](https://docs.ros.org/en/foxy/Tutorials/Intermediate/Writing-an-Action-Server-Client/Cpp.html#writing-the-action-server-code)

Reply:
Re-opening after thinking about it this weekend, `handeled_cancel` is much better

Reply:
Done in
https://github.com/tgroechel/rclcpp/pull/3/commits/b442bff40b7f9831352aede655934975a0ad9da4